### PR TITLE
Refactor javascript code - remove usage of event attributes

### PR DESCRIPTION
**Pull request by xmorave2**
TODO:
- Test templates:
  - [x] Helpers/email-form-fields.phtml
  - [x] Recommend/MapSelection.phtml
  - [x] Recommend/SideFacets/cluster-list.phtml
  - [x] Recommend/TopFacets.phtml 
  - [x] Recommend/WorldCatIdentities.phtml 
  - [x] RecordDriver/DefaultRecord/list-entry.phtml
  - [x] RecordDriver/SolrOverdrive/list-entry.phtml
  - [x] breadcrumbs/multi.phtml
  - [x] cart/cart.phtml
  - [x] devtools/language.phtml 
  - [x] install/showsql.phtml 
  - [x] myresearch/holds.phtml 
  - [x] myresearch/illrequests.phtml 
  - [x] myresearch/storageretrievalrequests.phtml 
  - [x] record/storageretrievalrequest.phtml 
  - [x] record/taglist.phtml
  - [x] search/advanced/layout.phtml
  - [x] search/reservessearch.phtml
  - [x] upgrade/fixanonymoustags.phtml 
  - [x] upgrade/fixduplicatetags.phtml 
  - [x] upgrade/fixmetadata.phtml
- [x] Run full automated test suite

### CODE REVIEW

**Comment by demiankatz** [comment]
Thanks again for this, @xmorave2; I'll try to help out with some of the template testing when time permits. In the meantime, if @crhallberg has any time to assist, that would be great, too!

**Comment by demiankatz** [comment]
I just ran the Mink-based automated tests and encountered some failures:

<pre>
There were 8 failures:

1) VuFindTest\Mink\CartTest::testFillAndDeleteFromCart
Selector not found: .modal-body .alert
Failed asserting that false is true.

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/MinkTestCase.php:246
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CartTest.php:239
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CartTest.php:434
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/AutoRetryTrait.php:86

2) VuFindTest\Mink\CartTest::testCartEmail
Selector not found: .modal-body .alert
Failed asserting that false is true.

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/MinkTestCase.php:246
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CartTest.php:239
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CartTest.php:550
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/AutoRetryTrait.php:86

3) VuFindTest\Mink\CartTest::testCartExport
Selector not found: .modal-body .alert
Failed asserting that false is true.

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/MinkTestCase.php:246
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CartTest.php:239
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CartTest.php:640
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/AutoRetryTrait.php:86

4) VuFindTest\Mink\CartTest::testCartPrint
Selector not found: .modal-body .alert
Failed asserting that false is true.

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/MinkTestCase.php:246
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CartTest.php:239
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/CartTest.php:719
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/AutoRetryTrait.php:86

5) VuFindTest\Mink\FavoritesTest::testAddRecordToFavoritesLogin
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'Invalid login -- please try again.'
+'We cannot log you in at this time. Please try again later.'

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/UserCreationTrait.php:84
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/FavoritesTest.php:187
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/AutoRetryTrait.php:86

6) VuFindTest\Mink\FavoritesTest::testAddRecordToFavoritesLoggedIn
Selector not found: #save_list
Failed asserting that false is true.

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/MinkTestCase.php:246
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/FavoritesTest.php:239
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/AutoRetryTrait.php:86

7) VuFindTest\Mink\FavoritesTest::testAddSearchItemToFavoritesLogin
Selector not found: #make-list
Failed asserting that false is true.

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/MinkTestCase.php:246
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/MinkTestCase.php:261
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/FavoritesTest.php:338
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/AutoRetryTrait.php:86

8) VuFindTest\Mink\FavoritesTest::testAddSearchItemToFavoritesLoggedIn
Selector not found: #save_list
Failed asserting that false is true.

/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/MinkTestCase.php:246
/home/dkatz/vufind3/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/FavoritesTest.php:383
/home/dkatz/vufind3/module/VuFind/src/VuFindTest/Unit/AutoRetryTrait.php:86

FAILURES!
Tests: 1373, Assertions: 5074, Failures: 8, Skipped: 1.
</pre>

I don't have time to troubleshoot these today, so I'm sharing them in case somebody else has time to investigate further. I imagine some may be legitimate problems while some may require the tests to be adjusted to account for template changes. I'll help more when time permits; I'm just having an unusually large number of long meetings this week, so my coding opportunities are limited.

**Comment by demiankatz** [comment]
Thanks for the fixes, @xmorave2; would you like me to try running the test suite again?

**Comment by xmorave2** [comment]
So I've done some testing and polishing things...

@demiankatz Could you guide me how to test the non-checked todo list item?
- ad MapSelection, I have some records with lat_long data and  MapSelection recommendation module enabled, but I can't see a map at the top of results

- ad SolrOverdrive/list-entry - I do not have acces to any solrdrive data, could you test it?

- ad cart and breadcrumbs/multi - I am not sure how to test? In which case should be the script executed?

- ad storageretrievalrequest - I was not able to enable it on top of Demo driver, is it even possible?

**Comment by demiankatz** [comment]
@xmorave2, thanks for the great progress here! I'll look into this in more depth a little later, but in the meantime, a few quick answers/comments:

- The MapSelection recommendation module is linked to settings in geofeatures.ini; I can't remember all the details, but it's possible that some additional configuration is necessary there to get things working. I can spend more time investigating it if this doesn't help.

- I don't have access to Overdrive either; I'll see whether I can figure out a way to simulate it, and if not, we can ask @bpalme for help.

- To access the cart functionality, you need to set showBookBag to true in config.ini; I believe that the Javascript is related to button processing in the cart pop-up.

- I'm pretty sure that breadcrumbs/multi is related to deprecated VuDL functionality and is never used. Perhaps @crhallberg can confirm, and we can remove the logic if it is not really needed. Right now I think it only serves to confuse people.

- Storage retrieval requests should work with the Demo driver, but sometimes it takes a few attempts to get a good example because of the random nature of the Demo driver data (simulating failures, etc.). I don't see any additional configuration that should be necessary, but maybe I'm forgetting something. I can definitely test this more thoroughly on my end if you're still having trouble on yours.

I'll comment again when I have enough time to actually try some of these things on my own; please don't feel any pressure to spend too much more of your own time if things are continuing to give you trouble. I just wanted to provide some pointers in case it helps get you unstuck while I'm occupied with other issues!

**Comment by xmorave2** [comment]
Thanks @demiankatz, I was able to enable MapSelection, and realize, that the code I was trying to change is not used, so I removed it.

**Comment by xmorave2** [comment]
I was able to test cart and record/storageretriavelrequest now, thanks for all suggestions @demiankatz

**Comment by demiankatz** [comment]
I have just repeated the test suite, and all automated tests are passing now. That's a good sign! I'm still going to do a bit more hands-on checking before merging, to be on the safe side.

**Comment by bpalme** [comment]
I'm sorry I haven't been able to test it out for Overdrive.  I don't anticipate being able to soon.  But I'll keep in on my to-do list.

Brent

**Comment by demiankatz** [comment]
@xmorave2, I've started merging chunks of this to master as I review them. I made a few adjustments to the DevTools/Language customizations, because the "extra" and "missing" buttons were rendering inconsistently. See https://github.com/vufind-org/vufind/commit/1b9ceeea1d0a043755cd422f1af305d8424a8a07 for the slightly revised version.

**Comment by demiankatz** [comment]
> I'm sorry I haven't been able to test it out for Overdrive. I don't anticipate being able to soon. But I'll keep in on my to-do list.

No worries, @bpalme! I'll do my best to test it by faking the data, and if you can review at some point in the future, we can address any outstanding problems. Worst case scenario, if you can test 7.0 when it is released, any necessary fixes can go into 7.0.1.

File: themes/bootstrap3/templates/Recommend/WorldCatIdentities.phtml [hunk 1]
```diff
@@ -14,17 +14,30 @@
...
       <?php endforeach; ?>
-      <?php if ($i > 3): ?><dd><a href="#" onclick="lessFacets('WCIdents'); return false;"><?=$this->transEsc('less')?> ...</a></dd></span><?php endif; ?>
+      <?php if ($i > 3): ?><dd id="less-WCIdents"><a href="#"><?=$this->transEsc('less')?> ...</a></dd></span><?php endif; ?>
     </dl>
   </div>
 <?php endif; ?>
+<?php
+$script = <<<JS
+$('#more-WCIdents a').click(function(e) {
+    moreFacets('WCIdents');
```
Comment (demiankatz) [inline themes/bootstrap3/templates/Recommend/WorldCatIdentities.phtml:34]: This template calls moreFacets/lessFacets, but these have been removed from common.js. Unless I'm missing something, I think that means there is a problem here.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/list-entry.phtml [hunk 2]
```diff
@@ -188,7 +188,7 @@
             <?=$this->transEsc('Delete') ?>
           </a>
           <ul class="dropdown-menu" role="menu" aria-labelledby="<?=$dLabel ?>">
-            <li><a onClick="$.post('<?=$deleteUrl?>', {'delete':'<?=$this->escapeJs($id) ?>','source':'<?=$this->escapeJs($source) ?>','confirm':true},function(){location.reload(true)})" title="<?=$this->transEsc('confirm_delete_brief')?>"><?=$this->transEsc('confirm_dialog_yes')?></a></li>
+            <li><a id="confirm_delete_item_<?=$this->escapeHtml($id);?>" title="<?=$this->transEsc('confirm_delete_brief')?>"><?=$this->transEsc('confirm_dialog_yes')?></a></li>
```
Comment (demiankatz) [inline themes/bootstrap3/templates/RecordDriver/DefaultRecord/list-entry.phtml:191]: Even with proper escaping, I'm concerned that there might be weird ID values that could create HTML-illegal IDs. How would you feel about using an md5 of the ID as the element ID, just to be on the safe side?

File: themes/bootstrap3/templates/Recommend/MapSelection.phtml [hunk 3]
```diff
@@ -24,10 +24,6 @@
     $searchParams = $this->recommend->getSearchParams();
     $coordinates = $this->recommend->getSelectedCoordinates();
 
-    if ($coordinates == null) {
```
Comment (demiankatz) [inline themes/bootstrap3/templates/Recommend/MapSelection.phtml:27]: How confident are you that it's safe to remove this logic? I agree that the `$addSearchOption` Javascript below seems to have no effect since I can find no references in the code to a `search-query-options` class, but the `$showSelection` and `$coordinates` values are being passed to the `loadMapSelection` function, and presumably have some effect.

I'm not very familiar with the geo-code, so I want to be sure we don't accidentally break something here. Perhaps @lmgonzales can shed some light if time permits.

**Comment by demiankatz** [comment]
Okay, I'm out of time for today, but I managed to merge a good chunk of the logic from this PR to master (and from there, back into here and #1494, reducing the outstanding diffs that need to be reviewed).

I left comments on a few things that I had concerns about; if @xmorave2 has time to answer those questions, I'd be grateful. I also left the facet refactoring as the last piece to review, since it's the most complicated, and I want to be sure it doesn't break @EreMaijala's ajax-loaded facets, etc. I'll test in more detail the next time I have a chance.

Thanks again for all of the work on this! I'll keep chipping away at all of this as time permits.

File: themes/bootstrap3/templates/Recommend/WorldCatIdentities.phtml (hunk 1 above)
Comment (xmorave2) [reply themes/bootstrap3/templates/Recommend/WorldCatIdentities.phtml]: I forget to fix this when I refactored facets js last time... now it should be fixed

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/list-entry.phtml (hunk 2 above)
Comment (xmorave2) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/list-entry.phtml]: I do think, it is not proper escaping, it should be `escapeHtmlAttr`, I tried to change, does it work for you now?

File: themes/bootstrap3/templates/Recommend/MapSelection.phtml (hunk 3 above)
Comment (xmorave2) [reply themes/bootstrap3/templates/Recommend/MapSelection.phtml:27]: I tried to experiment a bit with it and even at start of geographic searching, `getSelectedCoordinates` does not return `null` but default coordinates. And `showSelection` parameter is not consumed in `loadMapSelection` so I removed it.

**Comment by xmorave2** [comment]
Thanks a lot @demiankatz, I answer to your comments and also added few follow-ups.

File: themes/bootstrap3/templates/Recommend/MapSelection.phtml (hunk 3 above)
Comment (demiankatz) [reply themes/bootstrap3/templates/Recommend/MapSelection.phtml:27]: Yes, thank you, that makes sense. It appears that there may be some code paths where `getSelectedCoordinates` might be able to return a null value, so out of an abundance of caution, I included `getDefaultCoordinates` as a `??` fallback. It's possible that's not necessary, but it shouldn't hurt anything, and it still dramatically simplifies the code. Here's the final version I pushed to master: https://github.com/vufind-org/vufind/commit/051aae490ff21760b6deb55d2fd43c343939f05c

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/list-entry.phtml (hunk 2 above)
Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/list-entry.phtml]: Great, thanks! I was mainly concerned about very strange IDs breaking the logic, but based on [this](https://mathiasbynens.be/notes/html5-id-class) it sounds like that's probably not a major concern. :-)

File: themes/bootstrap3/templates/Recommend/WorldCatIdentities.phtml (hunk 1 above)
Comment (demiankatz) [reply themes/bootstrap3/templates/Recommend/WorldCatIdentities.phtml]: I think a little more work may be needed, for two reasons:

1.) If you load the WorldCatIdentities recommendation module on a page that has no other facet recommendation modules, facets.js won't load, and then the functionality won't work. It is probably worth including the JS file in the template; it will be automatically deduplicated, so I don't think there's any negative side effect to doing that.

2.) I haven't tested yet, but I'm concerned that calling registerMoreLessFacetsEventHandlers() multiple times may register duplicate event handlers, which might have weird side effects. It might be wise to `.unbind('click')` before setting up the event handlers.

If you don't have time for this, I can do more testing/adjustments tomorrow or Monday!

**Comment by demiankatz** [comment]
Thanks again, @xmorave2, for the continued progress; I got four more files merged to master today. I'll try to test the remaining logic further tomorrow or Monday, time permitting!

File: themes/bootstrap3/templates/Recommend/WorldCatIdentities.phtml (hunk 1 above)
Comment (xmorave2) [reply themes/bootstrap3/templates/Recommend/WorldCatIdentities.phtml]: Ad 1) it is the same with top facets, I added facets.js to both templates

Ad 2) yes, it is probably better to be on the safer side... register function now first remove the handler. As the handlers body is written now, it should have no side effects, but it could change in the future... we should not rely on that the implementation is always save and free of side effects...

**Comment by demiankatz** [comment]
Thanks for the finishing touches, @xmorave2. I've tested all the edge cases I could think of, and this seems fine; @crhallberg also gave the thumbs up. I'm running the automated tests one last time to be sure all is well, and assuming they pass, I'll merge this within the next few minutes.

**Comment by xmorave2** [comment]
thanks @demiankatz I appreciate your help here!
